### ISSUE TITLE
Verify provenance chain upon upload

### ISSUE THREAD

**Issue by autodidaddict** [wascc/gantry#3]
When attempting to upload a token to the gantry server, it needs to verify the provenance:
* For a `M`odule, verify that the `issuer` is among the list of `valid_signers` of a stored account, or is the public key of a stored account.
* For an `A`ccount, verify that the `issuer` is among the list of `valid_signers` embedded in the operator JWT. We should _not_ compare against the operator's root public key, because we don't want anyone to sign accounts with the operator's self-signing key.

**Comment by Spora-GH** [comment]
I'd like to have a go at this...

2 ways come to mind: 
set up a "provenance" trait with a bool function that module and account can implement, or inline the comparison (within fn upload, line 236 in gantry/main.rs) with a match statement to comparisons with either catalog::query_catalog(module) or claims.metadata.valid_signers (for accounts).

this would be a first contribution (for anything), so do let me know if i'm on the right path, or should take a different approach altogether...

cheers

**Comment by autodidaddict** [comment]
First, thanks for wanting to help out! The more contributors the merrier!

Here's a rough idea of the thoughts I had on provenance verification when I wrote up this issue:
* Every module/actor JWT has an `issuer` and a `subject` field. The `subject` field is the public key that uniquely identifies the actor, and the `subject` is the public key of the entity that _signed_ that actor's JWT.
* Every **account** has an `issuer` and a `subject` field, but, as you point out, it also has a `valid_signers` field in the claims metadata. 
* Every **operator** has an `issuer` and a `subject` field, and the claims metadata also contains a `valid_signers` field.

The rules for establishing provenance are:
* The `issuer` of an **Actor** must be within the `valid_signers` list of the **Account** that signed that actor's JWT, or it can optionally be the `subject` of the signing account.
* The `issuer` of an **Account** must be within the `valid_signers` list of the **Operator** that signed the Account's JWT. While this is an opinionated take on chained JWT verification, the issuer of an account _cannot_ be the public key of the operator, it _must only be_ one of the operator's `valid_signers`. This is to force the Gantry administrator to create a separate signing key for the operator to prevent creating new valid accounts from a stolen operator key.

In isolation, an Account or an Actor can't know their own provenance, so I would recommend against putting a trait or a function on the `Claims` themselves. In security terms, we need to keep a clean separation of the claims and the entities that verify the claims.

My original thinking for this was to modify the `catalog` actor so that it would validate the provenance of each JWT before insertion. This is why the `catalog` actor is given a configuration message that contains the root operator information. If you look at the code in that actor, you'll see it's given a `Vec<String>` of the operator signers (but not the operator public key, as discussed above).

In other words:
* Before the catalog actor allows the insertion of an Account, it checks that the `issuer` of the account is in the `OPERATOR_SIGNERS` vec. if not the account token is rejected.
* Before the catalog actor allows the insertion of an Actor, it checks to see if the `issuer` of the actor is one of the Accounts already residing in the catalog. If it is not, the actor token is rejected.

I would probably start by creating some kind of `verify_provenance` function inside the catalog.rs file, and invoke it somewhere in the [put_token](https://github.com/wascc/gantry/blob/master/catalog/src/catalog.rs#L6) function.

**Comment by autodidaddict** [comment]
Looking at the code again, I see that I've got the signature / expiration validation inside the `write_token` function, so maybe the call to verify the provenance should go somewhere around [here](https://github.com/wascc/gantry/blob/master/catalog/src/catalog.rs#L80), in `write_token`

**Comment by Spora-GH** [comment]
ok thanks, will let you know if any more questions. I sent you an email yesterday...

**Comment by autodidaddict** [comment]
I got it... sorry for the late reply, responding shortly.

**Comment by Spora-GH** [comment]
no worries

**Comment by autodidaddict** [comment]
I am doing some work on master right now, creating a manifest file, etc. It shouldn't impact you much, but warning you of the impending merge later today ;)

**Comment by Spora-GH** [comment]
all gd, will be starting on this tomorrow

**Comment by Spora-GH** [comment]
@autodidaddict  not sure where best to comment... are yesterday's changes a solution for now/has anything changed re: approach? apologies for yest... connection issues so I gave up after a while

**Comment by autodidaddict** [comment]
Yesterday's changes added a "find_issuer" function to the code which should maybe make the provenance validation easier to write. Requirements haven't changed. feel free to email me if you have more questions

**Comment by Spora-GH** [comment]
ok cool

**Comment by autodidaddict** [comment]
As of the latest version in master, you cannot put an account token that wasn't signed by one of the operator keys. You cannot put an actor token that wasn't signed by one of an existing account's signing keys.

The only check remaining is to check the root operator keys (runtime injected) prior to allowing the insertion of an operator.
